server/src/main/java/timely/store/iterators/TimeSeriesGroupingIterator.java

```diff
@@ -0,0 +1,309 @@
+            return true;
+        } else {
+            try {
+                refillBuffer();
+                if (series.size() == 0) {
+                    return false;
+                } else {
+                    seriesIterator = series.iterator();
+                    Pair<Key, Double> p = seriesIterator.next();
+                    topKey = p.getFirst();
+                    topValue = new Value(MetricAdapter.encodeValue(p.getSecond()));
+                    return true;
+                }
+            } catch (IOException e) {
+                throw new RuntimeException("Error filling buffer", e);
+            }
+        }
+    }
+
+    @Override
+    public void next() throws IOException {
+        LOG.trace("next()");
+    }
+
+    @Override
+    public void init(SortedKeyValueIterator<Key, Value> source, Map<String, String> options, IteratorEnvironment env)
+            throws IOException {
+        super.init(source, options, env);
+        String filterOption = options.getOrDefault(FILTER, null);
+        if (null == filterOption) {
+            throw new IllegalArgumentException("Window size must be specified.");
+        }
+        String[] split = filterOption.split(",");
+        filters = new Double[split.length];
+        for (int i = 0; i < split.length; i++) {
+            filters[i] = Double.parseDouble(split[i]);
+        }
+        LOG.trace("init - filter: {}", Arrays.toString(filters));
+    }
+
+    @Override
+    public void seek(Range range, Collection<ByteSequence> columnFamilies, boolean inclusive) throws IOException {
+        super.seek(range, columnFamilies, inclusive);
+        series.clear();
+        fillBuffer();
+        seriesIterator = series.iterator();
+    }
+
+    private void refillBuffer() throws IOException {
+        LOG.trace("refill()");
+        Metric top = null;
+        while (super.hasTop()) {
+            Key k = super.getTopKey();
+            Metric m = MetricAdapter.parse(k, super.getTopValue());
+            timely.model.Value v = m.getValue();
+            m.setValue(null);
+            Collections.sort(m.getTags());
+            if (top != null && top.equals(m)) {
+                // we have cycled back to the first series
+                break;
+            }
+            if (top == null) {
+                top = m;
+            }
+            TimeSeries values = series.get(m);
+            // new time series found, ignore it
+            if (null != values) {
+                LOG.trace("Adding value {} to series {}", v.getMeasure(), m);
+                values.put(k, v.getMeasure(), true);
+            } else {
+                LOG.trace("Ignoring new time series {}", m);
+            }
+            super.next();
+        }
+        removeSeriesWithMissingValues();
+    }
+
+    private void fillBuffer() throws IOException {
+        // fill buffer
+        Metric top = null;
+        int seenTop = 0;
+        while (super.hasTop()) {
```

Comment:
To deal with sparse metrics, this routine can be changed to call "refillBuffer" above (with suggested changes) until you have at least one metric that has been completely filled such that an "answer" can be computed.